language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -207,6 +181,132 @@ newtype DynamicBuiltinNameMeanings = DynamicBuiltinNameMeanings
     { unDynamicBuiltinNameMeanings :: Map DynamicBuiltinName DynamicBuiltinNameMeaning
     } deriving (Semigroup, Monoid)
 
+type Evaluator f m = DynamicBuiltinNameMeanings -> f TyName Name () -> m EvaluationResult
+
+type Evaluate m = ReaderT (Evaluator Term m) m
```

Comment:
It's monadic right now only because the CEK machine is monadic, but I believe the latter should throw exceptions. However if we manage to build the `CekT` transformer, having monadic evaluators will be useful of course. In general, the current design is fine, but we may want to revisit it in future.